README
====================================================
Python bindings for Amazon's Product Advertising API
====================================================

This module offers a light-weight access to the latest version of the Amazon
Product Advertising API without getting in your way. 

The `Amazon Product Advertising API`_ provides programmatic access to Amazon's
product selection and discovery functionality. It has search and look up
capabilities, provides information on products and other features such as
Reviews, Similar Products and New and Used listings.

.. _Amazon Product Advertising API: https://affiliate-program.amazon.com/gp/advertising/api/detail/main.html

Basic usage
===========

In order to use this API you'll obviously need an Amazon Associates Web Service
account for which you must with Amazon at http://aws.amazon.com. Each account
contains an *AWSAccessKeyId* and a *SecretKey*. 

Here is an example how to use the API to search for books of a certain 
publisher::

    AWS_KEY = '...'
    SECRET_KEY = '...'
    
    api = API(AWS_KEY, SECRET_KEY, 'de')
    node = api.item_search('Books', Publisher='Galileo Press')

The ``node`` object returned is a `lxml.objectified`_ element. All its 
attributes can be accessed the pythonic way::
    
    total_results = node.Items.TotalResults.pyval
    total_pages = node.Items.TotalPages.pyval
    
    # get all books from result set and 
    # print author and title
    for book in node.Items.Item:
        print '%s: "%s"' % (book.ItemAttributes.Author, 
                            book.ItemAttributes.Title)

Please refer to the more extensive `documentation`_ for more details.

.. _lxml.objectified: http://codespeak.net/lxml/objectify.html
.. _documentation: http://packages.python.org/python-amazon-product-api/

Status
======

This module is still undergoing development. All operations can be used with
the API's ``call(Operation=...)`` method. Additionally, there is a number of 
convenience methods that include for instance additional error checking.
These are currently limited to:
   
- ``ItemLookup``
- ``ItemSearch``
- ``SimilarityLookup``
- ``BrowseNodeLookup``

More functionality is to follow as development progresses. 

Installation
============

In order to install python-amazon-product-api you can use::

    pip install python-amazon-product-api
    
or download the source package from 
http://pypi.python.org/pypi/python-amazon-product-api, untar it and run ::
    
    python setup.py install

You'll also find binaries there to make your life easier if you happen to use
a Windows system. If not, please send me an e-mail and complain loudly!

Development
===========

The development version is available `bitbucket.org`_. Feel free to clone the 
repository and add your own features. ::
    
    hg clone http://bitbucket.org/basti/python-amazon-product-api/
    
Patches are always welcome! Please make sure that your change does not break 
the tests!

.. _bitbucket.org: http://bitbucket.org/basti/python-amazon-product-api/

License
=======

This module is release under the BSD License. You can find the full text of
the license in the LICENSE file.

Changelog
=========

.. include:: CHANGES


CHANGES
0.2.8 (2014-03-30)
------------------

- Fixed #31: Using generic error factory `_e()`, the original parsed XML can be
  accessed in case of failure.
- Added `InvalidAccount` and `InvalidSignature` exceptions (thanks to Jannis
  Gebauer)

0.2.7 (2013-10-08)
------------------

Small bugfix release!

- Examples fixed. Processors can now be processor instances again.

0.2.6 (2013-09-14) "Humperdinck"
--------------------------------

- Supports API version 2011-08-01
- Config files added
- Almost total rewrite of processors backend. Will work now with elementtree
  from stdlib, too.
- #26: Added endpoints for CN, ES and IT.
- Added RetryAPI to contrib package thanks to Jerry Ji.
- Documentation was overhauled.
- As of 2012-10-01 there are no more XSLT endpoints!

0.2.5 (2011-09-19) "Buttercup"
------------------------------

- Support for XSLT requests.
- Support for Associate tags thanks to Kilian Valkhof.
- New API versions 2010-12-01, 2010-11-01, 2010-10-01, 2010-09-01 and 2010-08-06
  added. 
- Fixed #16: Cannot install module under Python 2.4 without pycrypto being 
  installed first.
- `tox`_ (and `hudson`_) are now used for testing all supported Python versions
  (which includes Python 2.7 now, too).
- Test server is replaced with `pytest-localserver`_.
- Fixed #18: Throttling no longer block CPU (Thanks to Benoit C).
- Added response-caching API (in ``amazonproduct.contrib.caching``) to ease
  development (Thanks to Dmitry Chaplinsky for the idea).
- API explicitly warns about deprecated operations.

.. important:: The following operations are deprecated since 15 July 2010 and
   are now answered with a '410 Gone' (and a ``DeprecatedOperation`` exception):
   
   * CustomerContentLookup
   * CustomerContentSearch
   * Help
   * ListLookup
   * ListSearch
   * TagLookup
   * TransactionLookup
   * VehiclePartLookup
   * VehiclePartSearch
   * VehicleSearch

- Added new exceptions InvalidClientTokenId and MissingClientTokenId.
- ``REQUESTS_PER_SECONDS`` can now be floats as well (e.g. 2500/3600.0).
- Added test options options ``--api-version``, ``--locale`` and ``--refetch``.

.. _tox: http://codespeak.net/tox/
.. _hudson: http://jenkins.rotekroete.de/
.. _pytest-localserver: http://pypi.python.org/pypi/pytest-localserver/

0.2.4.1 (2010-06-23)
--------------------

Bugfix release! High time I get some continuous integration set up!

- Fixed #13: The module did not run under Python 2.4. Ooops!


0.2.4 (2010-06-13)
------------------

- Locale parameter is now required at initialisation. ::

      # before you could write
      api = API(AWS_KEY, SECRET_KEY)
      
      # now you have to specify your locale
      api = API(AWS_KEY, SECRET_KEY, 'de')

- Custom test server (``tests.server.TestServer``) added. It runs on localhost 
  and mimicks the Amazon webservice by replaying local XML files.
- Testing now supports multiple locales. Please not that you have to run 
  ``python setup.py test`` to run the unittests.
- ResultPaginator now also works with XPath expressions for attributes (Bug 
  reported Giacomo Lacava).
- Custom lookup for XML elements (during parsing) ensures that ``<ItemId/>``
  and ``<ASIN>`` are now always ``objectify.StringElement`` (Bug reported by 
  Brian Browning).
- Fixed #11: Module can now be installed library without lxml being installed 
  first.
- Regular expressions for parsing error messages can now deal with the Japanese
  version.

  .. warning:: The support for the Japanese locale (``jp``) is still very 
     experimental! A few error messages have still to be translated and the
     functionality has to be confirmed. If you know Japanese, get in touch!


0.2.3 (2010-03-20)
------------------

- Tests run now for all API versions. Test cases can now be told which versions
  to use (class attribute ``api_versions`` set to i.e. ``['2009-10-01']``).
- A custom AWS response processor can now be defined. For instance, here is one 
  using ``xml.minidom`` instead of ``lxml``::
      
      def minidom_response_parser(fp):
          root = parse(fp)
          # parse errors
          for error in root.getElementsByTagName('Error'):
              code = error.getElementsByTagName('Code')[0].firstChild.nodeValue
              msg = error.getElementsByTagName('Message')[0].firstChild.nodeValue
                  raise AWSError(code, msg)
              return root
          api = API(AWS_KEY, SECRET_KEY, processor=minidom_response_parser)
          root = api.item_lookup('0718155157')
          print root.toprettyxml()
          # ...
      
- Fixed #3: Support for API Version 2009-11-01.
- Fixed #4: When using a bad parameter combination, an 
  ``InvalidParameterCombination`` exception is raised. 
- Fixed #5: ``InvalidSearchIndex`` is raised when unknown ``SearchIndex`` is 
  specified.
- Fixed #7: Specifying API versions works now for more than just one test per 
  test case.
- The ``setup.py`` command has been empowered a bit with the following 
  additional options: ``test``, ``build_sphinx``, ``upload_sphinx``.
- ``ResultPaginator`` attributes ``_get_current_page_numer``, 
  ``_get_total_results`` and ``_get_total_page_numer`` are now private.


0.2.2 (2010-01-30)
------------------

- ``browse_node_lookup`` operation added.	
- ``help`` operation added.	
- ``list_lookup`` and ``list_search`` operations added.
- Default timeout for API calls is set to 5 sec.
- Test cases for correct parsing of XML responses added. Local XML files are
  used for testing (if available) stored in separate directories 
  according to API version. These can be overwritten when config value
  ``OVERWRITE_TESTS`` is set to ``True``.
- ``InvalidItemId`` exception is replaced by more general 
  ``InvalidParameterValue`` exception.	


0.2.1 (2009-11-20)
------------------

- Support for Python 2.4 added. 
- Fixed #2: ResultPaginator now returns ``None`` if the XPath expression doesn't 
  find the node it's looking for.
  

0.2.0 (2009-11-07) "Westley"
----------------------------


amazon/product.py

"""
Amazon Product Advertising API
==============================

    The Product Advertising API provides programmatic access to Amazon's
    product selection and discovery functionality so that developers like you
    can advertise Amazon products to monetize your website.
    
    The Product Advertising API helps you advertise Amazon products using
    product search and look up capability, product information and features
    such as Customer Reviews, Similar Products, Wish Lists and New and Used
    listings. You can make money using the Product Advertising API to advertise
    Amazon products in conjunction with the Amazon Associates program. Be sure
    to join the Amazon Associates program to earn up to 15% in referral fees
    when the users you refer to Amazon sites buy qualifying products.  

More info can be found at
https://affiliate-program.amazon.com/gp/advertising/api/detail/main.html

Requirements
------------

You need an Amazon Webservice account.

Kudos
-----

The ``_build_url()`` function is based on code by Adam Cox (found at 
http://blog.umlungu.co.uk/blog/2009/jul/12/pyaws-adding-request-authentication/)

"""

from base64 import b64encode
from datetime import datetime, timedelta
from hashlib import sha256
import hmac
from lxml import objectify
import re
from time import strftime, gmtime
from urlparse import urlsplit
from urllib2 import quote, urlopen

__docformat__ = "restructuredtext en"

LOCALES = {
    'ca' : 'http://ecs.amazonaws.ca/onca/xml', 
    'de' : 'http://ecs.amazonaws.de/onca/xml', 
    'fr' : 'http://ecs.amazonaws.fr/onca/xml', 
    'jp' : 'http://ecs.amazonaws.jp/onca/xml', 
    'uk' : 'http://ecs.amazonaws.co.uk/onca/xml', 
    'us' : 'http://ecs.amazonaws.com/onca/xml', 
}

class UnknownLocale (Exception):
    """
    Raised when unknown locale is specified.
    """
    
class AWSError (Exception):
    """
    Generic AWS error message.
    """
    def __init__(self, code, msg):
        self.code = code
        self.msg = msg
    def __str__(self):
        return '%(code)s: %(msg)s' % self.__dict__
    
class InvalidSearchIndex (Exception):
    """
    The value specified for SearchIndex is invalid. Valid values include:

    All, Apparel, Automotive, Baby, Beauty, Blended, Books, Classical, DVD,
    Electronics, ForeignBooks, HealthPersonalCare, HomeGarden, HomeImprovement,
    Jewelry, Kitchen, Magazines, MP3Downloads, Music, MusicTracks,
    OfficeProducts, OutdoorLiving, PCHardware, Photo, Shoes, Software,
    SoftwareVideoGames, SportingGoods, Tools, Toys, VHS, Video, VideoGames,
    Watches
    """
    
class InvalidResponseGroup (Exception):
    """
    The specified ResponseGroup parameter is invalid. Valid response groups for
    ItemLookup requests include:

    Accessories, AlternateVersions, BrowseNodes, Collections, EditorialReview,
    Images, ItemAttributes, ItemIds, Large, ListmaniaLists, Medium,
    MerchantItemAttributes, OfferFull, OfferListings, OfferSummary, Offers,
    PromotionDetails, PromotionSummary, PromotionalTag, RelatedItems, Request,
    Reviews, SalesRank, SearchBins, SearchInside, ShippingCharges,
    Similarities, Small, Subjects, Tags, TagsSummary, Tracks, VariationImages,
    VariationMatrix, VariationMinimum, VariationOffers, VariationSummary,
    Variations.
    """

class InvalidItemId (Exception):
    """
    The specified ItemId parameter is invalid. Please change this value and 
    retry your request.
    """

class NoSimilarityForASIN (Exception):
    """
    When you specify multiple items, it is possible for there to be no 
    intersection of similar items.
    """
    
INVALID_SEARCH_INDEX_REG = re.compile(
    'The value you specified for SearchIndex is invalid.')

INVALID_ITEMID_REG = re.compile('(.+?) is not a valid value for ItemId. '
    'Please change this value and retry your request.')

NOSIMILARITIES_REG = re.compile('There are no similar items for this ASIN: '
                                '(?P<ASIN>\w+).')

class API (object):
    
    """
    Wrapper class for the Amazon Product Advertising API. You will need both an 
    AWS access key id and the secret counterpart.
    
    Example::
    
        AWS_KEY = '...'
        SECRET_KEY = '...'
        
        api = ProductAdvertisingAPI(AWS_KEY, SECRET_KEY)
        root = api.item_lookup('987311264224', IdType='ISBN', 
                    SearchIndex='Books', ResponseGroup='Reviews', ReviewPage=1)
        
        rating = root.Items.Item.CustomerReviews.AverageRating.pyval
        total_reviews = root.Items.Item.CustomerReviews.TotalReviews.pyval
        review_pages = root.Items.Item.CustomerReviews.TotalReviewPages.pyval

    """
    
    VERSION = '2009-07-01' #: supported Amazon API version
    REQUESTS_PER_SECOND = 2 #: max requests per second
    
    def __init__(self, access_key_id, secret_access_key, locale='de'):
        
        self.access_key = access_key_id
        self.secret_key = secret_access_key
        
        try:
            parts = urlsplit(LOCALES[locale])
            self.scheme = parts.scheme
            self.host = parts.netloc
            self.path = parts.path
        except KeyError:
            raise UnknownLocale(locale)
        
        self.last_call = datetime(1970, 1, 1)
        self.throttle = timedelta(seconds=1)/self.REQUESTS_PER_SECOND
        
    def _build_url(self, **qargs):
        """
        Builds a signed URL for querying Amazon AWS.
        """
        
        if 'AWSAccessKeyId' not in qargs:
            qargs['AWSAccessKeyId'] = self.access_key
            
        if 'Service' not in qargs:
            qargs['Service'] = 'AWSECommerceService'
            
        # use the version this class was build for by default
        if 'Version' not in qargs:
            qargs['Version'] = self.VERSION
            
        # add timestamp (this is required when using a signature)
        qargs['Timestamp'] = strftime("%Y-%m-%dT%H:%M:%SZ", gmtime())
        
        # create signature
        keys = sorted(qargs.keys())
        args = '&'.join('%s=%s' % (key, quote(str(qargs[key]))) 
                        for key in keys)
        
        msg = 'GET'
        msg += '\n' + self.host
        msg += '\n' + self.path
        msg += '\n' + args.encode('utf-8')
        
        signature = quote(
            b64encode(hmac.new(self.secret_key, msg, sha256).digest()))
        
        url = '%s://%s%s?%s&Signature=%s' % (self.scheme, self.host, self.path, 
                                             args, signature)
        return url
    
    def _call(self, url):
        """
        Calls the Amazon Product Advertising API and objectifies the response.
        """
        # Be nice and wait for some time 
        # before submitting the next request
        while (datetime.now() - self.last_call) < self.throttle: 
            pass # Wait for it!
        self.last_call = datetime.now()
        
        tree = objectify.parse(urlopen(url))
        root = tree.getroot()
        
        #~ from lxml import etree
        #~ print etree.tostring(tree, pretty_print=True)
        
        nspace = root.nsmap.get(None, '')
        errors = root.xpath('//aws:Request/aws:Errors/aws:Error', 
                         namespaces={'aws' : nspace})
        
        for error in errors:
            raise AWSError(error.Code.text, error.Message.text)
        
        #~ from lxml import etree
        #~ print etree.tostring(root, pretty_print=True)
        return root
    
    def item_lookup(self, id, **params):
        """
        Given an Item identifier, the ``ItemLookup`` operation returns some or
        all of the item attributes, depending on the response group specified
        in the request. By default, ``ItemLookup`` returns an item's ``ASIN``,
        ``DetailPageURL``, ``Manufacturer``, ``ProductGroup``, and ``Title`` of
        the item.
        
        ``ItemLookup`` supports many response groups, so you can retrieve many
        different kinds of product information, called item attributes,
        including product reviews, variations, similar products, pricing,
        availability, images of products, accessories, and other information.
        
        To look up more than one item at a time, separate the item identifiers
        by commas. 
        """
        try:
            url = self._build_url(Operation='ItemLookup', ItemId=id, **params)
            return self._call(url)
        except AWSError, e:
            
            if (e.code=='AWS.InvalidEnumeratedParameter' 
            and INVALID_SEARCH_INDEX_REG.search(e.msg)):
                raise InvalidSearchIndex(params.get('SearchIndex'))
            
            if e.code=='AWS.InvalidResponseGroup': 
                raise InvalidResponseGroup(params.get('ResponseGroup'))
            
            if (e.code=='AWS.InvalidParameterValue' 
            and INVALID_ITEMID_REG.search(e.msg)):
                raise InvalidItemId(id)
            
            # otherwise re-raise exception
            raise
        
    def item_search(self, search_index, **params):
        """
        The ``ItemSearch`` operation returns items that satisfy the search
        criteria, including one or more search indices.

        ``ItemSearch`` returns up to ten search results at a time. When
        ``condition`` equals "All," ``ItemSearch`` returns up to three offers
        per condition (if they exist), for example, three new, three used,
        three refurbished, and three collectible items. Or, for example, if
        there are no collectible or refurbished offers, ``ItemSearch`` returns
        three new and three used offers.

        Because there are thousands of items in each search index,
        ``ItemSearch`` requires that you specify the value for at least one
        parameter in addition to a search index. The additional parameter value
        must reference items within the specified search index. For example,
        you might specify a browse node (BrowseNode is an ``ItemSearch``
        parameter), Harry Potter Books, within the Books product category. You
        would not get results, for example, if you specified the search index
        to be Automotive and the browse node to be Harry Potter Books. In this
        case, the parameter value is not associated with the search index
        value.

        The ``ItemPage`` parameter enables you to return a specified page of
        results. The maximum ``ItemPage`` number that can be returned is 400.
        An error is returned if you try to access higher numbered pages. If you
        do not include ``ItemPage`` in your request, the first page will be
        returned by default. There can be up to ten items per page.

        ``ItemSearch`` is the operation that is used most often in requests. In
        general, when trying to find an item for sale, you use this operation.
        """
        try:
            url = self._build_url(Operation='ItemSearch', 
                                  SearchIndex=search_index, **params)
            return self._call(url)
        except AWSError, e:
            
            if (e.code=='AWS.InvalidEnumeratedParameter' 
            and INVALID_SEARCH_INDEX_REG.search(e.msg)):
                raise InvalidSearchIndex(search_index)
            
            if e.code=='AWS.InvalidResponseGroup': 
                raise InvalidResponseGroup(params.get('ResponseGroup'))
            
            # otherwise re-raise exception
            raise
        
    def similarity_lookup(self, *ids, **params):
        """
        The ``SimilarityLookup`` operation returns up to ten products per page
        that are similar to one or more items specified in the request. This
        operation is typically used to pique a customer's interest in buying
        something similar to what they've already ordered.

        If you specify more than one item, ``SimilarityLookup`` returns the
        intersection of similar items each item would return separately.
        Alternatively, you can use the ``SimilarityType`` parameter to return
        the union of items that are similar to any of the specified items. A
        maximum of ten similar items are returned; the operation does not
        return additional pages of similar items. If there are more than ten
        similar items, running the same request can result in different answers
        because the ten that are included in the response are picked randomly.
        The results are picked randomly only when you specify multiple items
        and the results include more than ten similar items. 
        """
        item_id = ','.join(ids)
        try:
            url = self._build_url(Operation='SimilarityLookup', 
                                  ItemId=item_id, **params)
            return self._call(url)
        except AWSError, e:
            
            if e.code=='AWS.ECommerceService.NoSimilarities':
                asin = NOSIMILARITIES_REG.search(e.msg).group('ASIN')
                raise NoSimilarityForASIN(asin)

class ResultPaginator (object):
    
    """
    Wrapper class for paginated results. This class will call the passed 
    function iteratively until either the specified limit is reached or all 
    result pages are fetched.
    
    A small example fetching reviews for a book::
        
        api = API(AWS_KEY, SECRET_KEY)
        paginator = ResultPaginator('ReviewPage',
            '//aws:Items/aws:Request/aws:ItemLookupRequest/aws:ReviewPage',
            '//aws:Items/aws:Item/aws:CustomerReviews/aws:TotalReviewPages',
            '//aws:Items/aws:Item/aws:CustomerReviews/aws:TotalReviews')
        
        for root in paginator(api.item_lookup, id=isbn, IdType='ISBN', 
                             SearchIndex='Books', ResponseGroup='Reviews'):
            ...
    
    """
    
    def __init__(self, counter, current_page, total_pages, total_results, 
                 limit=None, nspace=None):
        """
        :param counter: counter variable passed to AWS.
        :param current_page: XPath expression locating current paginator page.
        :param total_pages: XPath expression locating total number of pages.
        :param total_results: XPath expression locating total number of results. 
        :param limit: limit fetched pages to this amount. 
        :param nspace: used XML name space. 
        """
        self.counter = counter
        self.current_page_xpath = current_page
        self.total_pages_xpath = total_pages
        self.total_results_xpath = total_results
        
        self.limit = limit
        self.nspace = nspace
        
    def __call__(self, fun, **kwargs):
        """
        Iterate over all paginated results of ``fun``.
        """
        current_page = 0
        total_pages = 1
        
        kwargs[self.counter] = kwargs.get(self.counter, 1)
        
        while (current_page < total_pages 
        and (self.limit is None or current_page < self.limit)):
            
            root = fun(**kwargs)
            
            if self.nspace is None:
                self.nspace = root.nsmap.get(None, '')
        
            current_page = self.get_current_page_numer(root)
            total_results = self.get_total_results(root)
            total_pages = self.get_total_page_numer(root)
            
            yield root
            
            kwargs[self.counter] += 1
        
    def get_total_page_numer(self, root):
        """
        Get total number of paginator pages.
        """
        return root.xpath(self.total_pages_xpath, 
                          namespaces={'aws' : self.nspace})[0].pyval
        
    def get_current_page_numer(self, root):
        """
        Get number of current paginator page.
        """
        return root.xpath(self.current_page_xpath, 
                          namespaces={'aws' : self.nspace})[0].pyval
    
    def get_total_results(self, root):
        """
        Get number of current paginator page.
        """
        return root.xpath(self.total_results_xpath, 
                          namespaces={'aws' : self.nspace})[0].pyval

